alf/algorithms/muzero_algorithm.py

```diff
@@ -295,70 +295,177 @@ def _hook(grad, name):
+            # [B, 1, 1]
+            env_ids = folded_env_ids.unsqueeze(-1)
 
             # [B]
+            start_positions = convert_device(batch_info.positions)
+
+            # [B, T + R], capped at the end of the replay buffer.
+            folded_positions = torch.min(
+                start_positions.unsqueeze(-1) + torch.arange(T + R),
+                replay_buffer.get_current_position()[start_env_ids, None] - 1)
+
+            # [B, T, R + 1]
+            positions = folded_positions.unfold(1, R + 1, 1)
+
+            # [B, T]
             steps_to_episode_end = replay_buffer.steps_to_episode_end(
-                positions, env_ids)
-            # [B]
-            episode_end_positions = positions + steps_to_episode_end
+                positions[:, :, 0], env_ids[:, :, 0])
 
-            # [B, unroll_steps+1]
-            positions = positions.unsqueeze(-1) + torch.arange(
-                self._num_unroll_steps + 1)
-            # [B, 1]
-            env_ids = batch_info.env_ids.unsqueeze(-1)
-            # [B, 1]
+            # [B, T]
+            episode_end_positions = positions[:, :, 0] + steps_to_episode_end
+            # [B, T, 1]
             episode_end_positions = episode_end_positions.unsqueeze(-1)
 
+            # [B, T, R + 1]
             beyond_episode_end = positions > episode_end_positions
             positions = torch.min(positions, episode_end_positions)
 
+            # 1. Some times we would like to make sure that a sequence (e.g.
+            #    reanalyze unrolled sequence) to stop at the episode end, and
+            #    elements in the sequence beyond the episode end is filled with
+            #    the value of the the episode end.
+            #
+            # 2. After the unfolding, the original sequence along dim=1 is
+            #    unfolded into T sequences of length R. If the original sequence
+            #    cross the boundaries of the episodes, the new unfolded
+            #    sequences may start at different episodes, and may therefore
+            #    have different episode ends.
+            #
+            # With the above in mind, the helper function below achieves
+            # unfolding and adapting on the episode ends individually for each
+            # unfolded sequences.
+
+            # Slicing ``capped_index`` is used to create a view of shape ([B, T,
+            # R + 1]) but each sequence (b, t) is capped by the index of the
+            # episode end if the sequence crosses the episode boundary.
+            capped_index = (
+                torch.arange(B)[:, None, None],  # [B, 1, 1]
+                torch.arange(T)[:, None],  # [T, 1]
+                torch.min(
+                    steps_to_episode_end.unsqueeze(-1),
+                    torch.arange(R + 1)))  # [B, T, 1]
+
+            def _unfold_dim1_adapting_episode_ends(x):
+                # Do the unfolding first.
+                return _unfold_dim1(x)[capped_index]
```

Comment:
As the comment in the previous review suggested, `_unfold_dim1` and `[capped_index]` can be merged as one single indexing operation.
```python
capped_index = (
        torch.arange(B)[:, None, None],  # [B, 1, 1]
        torch.arange(T)[:, None] + torch.min(
            steps_to_episode_end.unsqueeze(-1),
            torch.arange(R + 1)))  # [B, T, R+1]
def _unfold_dim1_adapting_episode_ends(x):
        return x[capped_index]
```